File: Medium/78. Subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> cur;
        vector<vector<int>> ans;
        for(int i = 0; i <= nums.size(); i++){
            combi(nums, i, 0, cur, ans);
        }
        return ans;
    }
private:
    void combi(vector<int>& nums, int n, int s, vector<int> &cur, vector<vector<int>> &ans){
        if(n == cur.size()){
            ans.push_back(cur);
            return;
        }
        for(int i = s; i < nums.size(); i++){
            cur.push_back(nums[i]);
            combi(nums, n, i + 1, cur, ans);
            cur.pop_back();
        }
    }
};

File: Medium/78. Subsets/subsets.java
class Solution {  
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> track = new ArrayList<>();
        backtrack(nums, 0, track, ans);
        return ans;
    }
    private void backtrack(int [] nums, int start, List<Integer> track, List<List<Integer>> ans){
        ans.add(new ArrayList<Integer>(track));
        for(int i = start; i < nums.length; i++){
            track.add(nums[i]);
            backtrack(nums, i + 1, track, ans);
            track.remove(track.size() - 1);
        }
    }
}